File: apps/api/app/services/investor_service.py
from psycopg import Connection

from app.repositories.investor_repository import investor_repository
from app.schemas.investor import InvestorDetail, InvestorListData


class InvestorService:
    def __init__(self, repository=investor_repository) -> None:
        self._repository = repository

    def list_summaries(self, connection: Connection) -> InvestorListData:
        items = self._repository.list_summaries(connection)
        sorted_items = sorted(items, key=lambda investor: investor.name)
        return InvestorListData(items=sorted_items)

    def get_detail(self, connection: Connection, slug: str) -> InvestorDetail | None:
        return self._repository.get_detail(connection, slug)


investor_service = InvestorService()

File: apps/web/src/features/investors/server/services/investor-service.ts
import "server-only";

import { InvestorError } from "@/features/investors/server/errors";
import { investorRepository } from "@/features/investors/server/repositories/investor-repository";
import {
  investorListDataSchema,
  type InvestorDetail,
  type InvestorListData,
} from "@/features/investors/types/investor";

export class InvestorService {
  async listSummaries(): Promise<InvestorListData> {
    const items = await investorRepository.listSummaries();
    const sortedItems = [...items].sort((left, right) =>
      left.name.localeCompare(right.name),
    );

    return investorListDataSchema.parse({ items: sortedItems });
  }

  async getDetail(slug: string): Promise<InvestorDetail> {
    const investor = await investorRepository.getDetail(slug);
    if (!investor) {
      throw new InvestorError("INVESTOR_NOT_FOUND");
    }

    return investor;
  }
}

export const investorService = new InvestorService();
